cf-oscore/src/main/java/org/eclipse/californium/oscore/OscoreOptionDecoder.java

```diff
@@ -85,26 +100,32 @@ private void decode() {
 		byte[] kidContext = null;
 		int index = 1;
 
-		// Parsing Partial IV
-		if (n > 0) {
-			partialIV = Arrays.copyOfRange(total, index, index + n);
-			index += n;
-		}
+		// Attempt to parse the bytes of the option
+		try {
+			// Parsing Partial IV
+			if (n > 0) {
+				partialIV = Arrays.copyOfRange(total, index, index + n);
+				index += n;
+			}
 
-		// Parsing KID Context
-		if (h != 0) {
-			int s = total[index++];
+			// Parsing KID Context
+			if (h != 0) {
+				int s = total[index++];
 
-			kidContext = Arrays.copyOfRange(total, index, index + s);
+				kidContext = Arrays.copyOfRange(total, index, index + s);
 
-			index += s;
-		}
+				index += s;
+			}
 
-		// Parsing KID
-		kid = Arrays.copyOfRange(total, index, total.length);
+			// Parsing KID
+			if (k != 0) {
+				kid = Arrays.copyOfRange(total, index, total.length);
+			}
```

Comment:
Yes, true, that is an extra check that can be done. (Additional checks will also be done later in decryptor.decodeObjectSecurity).

I now throw an Exception if k == 0 but index != total.length, now I just throw a simple exception but perhaps that is fine since it will be caught right below and thrown as CoapOSException with an error message. Updated in commit 210590ce8febf31324d5cd7400a68e6ec94efc01

Reply:
I updated the error reporting when parsing the OSCORE option to be a bit more fine-grained in commit 51523ef232437132857844087de5c19526163421